fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/ConcatCursor.java

```diff
@@ -0,0 +1,274 @@
+            }
+        } else {
+            //no continuation, start at the beginning of the first cursor
+            this.firstCursor = firstFunction.apply(context, scanProperties, currentCursorContinuation);
+            this.secondCursor = null;
+            this.currentCursorContinuation = null;
+        }
+    }
+
+    @Nonnull
+    @Override
+    public CompletableFuture<RecordCursorResult<T>> onNext() {
+        if (secondCursor == null) {
+            return firstCursor.onNext().thenCompose(result -> {
+                if (result.hasNext() || result.getNoNextReason().isLimitReached() || result.getNoNextReason().isOutOfBand()) {
+                    // switch to second cursor if a) no more rows b) the row limit is not yet reached and c) out-of-band limits not exceeded
+                    return CompletableFuture.completedFuture(result);
+                } else {
+                    Function<ExecuteProperties, ExecuteProperties> f = e -> e.toBuilder().setReturnedRowLimit(rowLimit).build(); //enforce row limit
+                    secondCursor = secondFunction.apply(context, this.scanProperties.with(f), null); //start at beginning of second cursor
+                    return secondCursor.onNext();
+                }
+            }).thenApply(this::postProcess);
+        } else {
+            return secondCursor.onNext().thenApply(this::postProcess);
+        }
+    }
+
+    // return concat result from underlying cursor result
+    private RecordCursorResult<T> postProcess(RecordCursorResult<T> result) {
+        nextResult = getConcatResult(result);
+        mayGetContinuation = !result.hasNext(); // shim to support old continuation style
+        return nextResult;
+    }
+
+    //if haven't exhausted both cursors and no limits reached, form a result that encapsulates the value returned by
+    //the current underlying cursor, plus a continuation that allows us to restart at same cursor at its next position
+    private RecordCursorResult<T> getConcatResult(RecordCursorResult<T> nextResult) {
+        @Nonnull RecordCursorResult<T> concatResult;
+        if (!nextResult.hasNext()) {
+            if (secondCursor != null && nextResult.getNoNextReason().isSourceExhausted()) {
+                concatResult = RecordCursorResult.exhausted(); //continuation not valid here
+            } else {
+                concatResult = RecordCursorResult.withoutNextValue(new ConcatCursorContinuation(secondCursor != null ? true : false, nextResult), nextResult.getNoNextReason());
+            }
+        } else {
+            concatResult = RecordCursorResult.withNextValue(nextResult.get(), new ConcatCursorContinuation(secondCursor != null ? true : false, nextResult));
+            rowLimit = Integer.max(0, --rowLimit);
+        }
+        return concatResult;
+    }
+
+    //form a continuation that allows us to restart with the current cursor at its next position
+    private class ConcatCursorContinuation implements RecordCursorContinuation {
```

Comment:
It is a little awkward having this in the middle of the outer class.  It would probably be better at the end of the outer class.